src/Sulu/Component/Content/Document/Syncronization/DocumentRegistrator.php

```diff
@@ -0,0 +1,214 @@
+use Sulu\Component\DocumentManager\Behavior\Mapping\ParentBehavior;
+use Sulu\Component\DocumentManager\DocumentManagerInterface;
+
+/**
+ * Class responsible for registering a document from the default document manager
+ * with the publish document manager including any documents dependent on
+ * the document to be synchronized.
+ */
+class DocumentRegistrator
+{
+    /**
+     * @var DocumentManagerInterface
+     */
+    private $defaultManager;
+
+    /**
+     * @var DocumentManagerInterface
+     */
+    private $publishManager;
+
+    public function __construct(
+        DocumentManagerInterface $defaultManager,
+        DocumentManagerInterface $publishManager
+    ) {
+        $this->defaultManager = $defaultManager;
+        $this->publishManager = $publishManager;
+    }
+
+    /**
+     * Register the incoming DDM document with any existing PHPCR node in the
+     * PDM.
+     *
+     * If the PDM already has the incoming PHPCR node then we need to register
+     * the existing PHPCR node from the PDM PHPCR session with the incoming DDM
+     * document (otherwise the system will attempt to create a new document and
+     * fail).
+     *
+     * We also ensure that any immediately related documents are also registered.
+     *
+     * @param SynchronizeBehavior $document
+     */
+    public function registerDocumentWithPDM(SynchronizeBehavior $document)
+    {
+        $this->registerSingleDocumentWithPDM($document);
+
+        $metadata = $this->defaultManager->getMetadataFactory()->getMetadataForClass(get_class($document));
+        // iterate over the field mappings for the document, if they resolve to
+        // an object then try and register it with the PDM.
+        foreach (array_keys($metadata->getFieldMappings()) as $field) {
+            $propertyValue = $metadata->getFieldValue($document, $field);
+
+            if (false === is_object($propertyValue)) {
+                continue;
+            }
+
+            // if the default document manager does not have this object then it is
+            // not a candidate for being persisted (e.g. it might be a \DateTime
+            // object).
+            if (false === $this->defaultManager->getRegistry()->hasDocument($propertyValue)) {
+                continue;
+            }
+
+            $this->registerSingleDocumentWithPDM($propertyValue);
+        }
+
+        // TODO: Workaround for the fact that "parent" is not in the metadata,
+        // see: https://github.com/sulu-io/sulu-document-manager/issues/67
+        if ($document instanceof ParentBehavior) {
+            if ($parent = $document->getParent()) {
+                $this->registerSingleDocumentWithPDM($parent, true);
+            }
+        }
+    }
+
+    private function registerSingleDocumentWithPDM($document, $create = false)
+    {
+        $ddmInspector = $this->defaultManager->getInspector();
+        $pdmRegistry = $this->publishManager->getRegistry();
```

Comment:
Then I would still say that `defaultInspector` and `publishedRegistry` are better.